# The-Specificity-Landscape-of-AtWRKYs-Reveals-the-Bidirectional-Influence-of-non-CG-Methylation
A distinct epigenetic feature of plants is the occurrence of DNA methylation in non-CG contexts. Although the physiological roles of non-CG methylation were elucidated, its direct impact on TF–DNA interactions remains largely unexplored. Focusing on WRKY-family TFs, here we investigated how non-CG methylation influences their DNA-binding specificity and genome-wide cis-regulatory elements (CREs). By generating 461 SELEX and DAP-seq libraries for 54 AtWRKYs, we show that DNA methylation alters both monomeric and dimeric binding specificities of WRKYs, leading to an overall increase in specificity divergence among family members. We curated 201 WRKY motifs and clustered them into 11 classes, 5 of which represent previously unreported specificities. Notably, the known WRKY cis-element PRE4 was found to be recognized only when methylated. The comprehensive dataset of accurate WRKY motifs also enabled the identification of the amino-acid discriminants of W-box and WT-box. Expanding on prior knowledge, we demonstrate that methylation not only decreases but can also increase WRKYs’ affinity. This bidirectional effect has globally reshaped the genomic binding landscape of WRKYs upon methylation. Finally, we constructed the WRKY Regulatory Code Database (https://transysbio.cn/WRKYRCDB.php) to facilitate data access.
The code for generating the main figures of the manuscript is provided here.

![image](https://github.com/Jiang-Bio/WRKY_RCDB/blob/master/graphical_abstract.png)
